Working Login

Allcharles opened the pull request:
# Working login with baw-client

Updated login method to be compatible with baw-client

## Changes

- Re-implemented login system

## Problems

- Solution is fairly hacky because of https://github.com/QutEcoacoustics/baw-server/issues/509

## Final Checklist

- [ ] Assign reviewers if you have permission
- [ ] Assign labels if you have permission
- [ ] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Cleanup app styling to create full height components

- Cleanup CSS styling for the app component so that the default router outlet component will have access to the full height of the page
- Created basic audio analysis page

Allcharles committed:
Working example

Couple bugs to fix
- AngularJS links are not reloading the parent page
- IFrame does not resize to content which causes the scrollbar to be in the wrong location

Allcharles committed:
Created new audio analysis page

Allcharles committed:
Cleanup iframe module

- Disabled broken tests
- Renamed IFrameComponent to BawClientComponent
- Isolated BawClientComponent into a module
- Created placeholder unit tests
- Updated interceptor to include cookies in requests

Allcharles committed:
Implemented BawClientComponent unit tests

Allcharles committed:
Fixed broken tests and rebase conflicts

Allcharles committed:
Fixed broken unit tests

Discovered bug in how secondary menu unit tests were saving/restoring the defaultMenu global variable (converted Lists to arrays, but not back)

Allcharles committed:
Fix site/region details tag links

Allcharles committed:
Working login request

Allcharles committed:
Update security service tests

Allcharles committed:
Cleanup interceptor tests

atruskie commented on src/app/services/baw-api/api.interceptor.service.ts:
```diff
@@ -46,7 +46,7 @@ export class BawApiInterceptor implements HttpInterceptor {
 
-    // Don't add these headers to requests to cms service
+    // Don't add these headers to requests to non json requests
     if (request.responseType !== "text") {
```
the condition does not match the comment. `aplication/octet-stream` is not text... should it have these headers?

atruskie replied:
I see this is part of the responseType type's definition:

```
  responseType: 'arraybuffer' | 'blob' | 'json' | 'text'
```

Still could clear up conditon/comment

atruskie commented on src/app/services/baw-api/baw-api.service.ts:
```diff
@@ -309,6 +309,15 @@ export abstract class BawApiService<Model extends AbstractModel> {
 
+  /**
+   * Concatenates path with apiRoot to form a full URL.
+   *
```
add comment on whether `path` is expected to have prefixed `/` or not

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
+      // Catch wrong type of response
+      if (typeof page !== "string") {
+        throwError(err);
+      }
```
Is this defensive or does this actually happen?

Allcharles replied:
Defensive

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
+
+      // Extract auth token if exists
+      const token = page.match(/name="authenticity_token" value="(.+?)"/);
+      if (typeof token[1] !== "string") {
```
a truthy check is probably sufficent here? or a check for null/undefined? regex match will always return a string

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -1,19 +1,24 @@
+function timestamp(x: number) {
+  return x;
+}
+
```
this seems wholly redundant

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
+      .pipe(
+        // Extract form data from login form
+        tap((page: string) => setFormData(page, formData)),
+        // Request api cookie
```
```suggestion
        // Mimic a traditional form-based sign in the get a well-formed auth cookie
        // Needed bause of https://github.com/QutEcoacoustics/baw-server/issues/509
```

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -58,7 +63,58 @@ export class SecurityService extends BawApiService<SessionUser> {
+        tap((page: string) => setFormData(page, formData)),
+        // Request api cookie
+        mergeMap(() => this.requestApiCookie(formData)),
+        // Get session user
```
```suggestion
        // Trade the cookie for an API auth token (mimicking old baw-client)
```

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -74,6 +130,21 @@ export class SecurityService extends BawApiService<SessionUser> {
+  /**
+   * Request api auth cookie
+   */
+  private requestApiCookie(formData: URLSearchParams) {
```
```suggestion
  private signInWithFormData(formData: URLSearchParams) {
```

atruskie commented on src/app/services/baw-api/security/security.service.ts:
```diff
@@ -74,6 +130,21 @@ export class SecurityService extends BawApiService<SessionUser> {
   }
 
+  /**
+   * Request api auth cookie
```
```suggestion
   * Use the form-based sign in to authenticate. The server will issue a traditional session cookie which we will later trade for an auth token.
```

atruskie commented on src/app/test/fakes/ApiErrorDetails.ts:
```diff
@@ -40,3 +40,22 @@ export function generateApiErrorDetails(
   };
 }
+
+export function generateApiErrorResponse(
```
```suggestion
/*
 * Used to return a well constructed error object for API responses that do not conform to the standard
 */
export function generateApiErrorResponse(
```

atruskie reviewed (approved):
Great work. Only minor suggestions for comments recommended

Allcharles committed:
Fix rebase issues

Allcharles committed:
Cleanup api request responseType's

This should make things a lot more clear for when the interceptor should set accept/content-type headers to requests

Allcharles committed:
Cleanup review comments
